### PR TITLE
Hero list feature API and ConnectedComponent

### PR DESCRIPTION
**Pull request by vinnyjth**
### What does this PR do, or why is it needed?
Adds the API for the HeroListFeature and the associated connected component. 

The component is driven with the following logic. 

If `heroAlgorithms` are provided, the first result from those algorithms drives the hero card. If an algorithm is not provided, the first entry off the `algorithms` array is sliced off and used as the action, and all other actions are used for the action list items underneath. 

Besides that, nothing special. 

Screen Shot 2020-04-29 at 9 10 15 PM
Screen Shot 2020-04-29 at 9 05 45 PM

### What design trade-offs/decisions were made?
n/a

### How do I test this PR?
You'll want to add something like this to your config.yml HOME_FEATURES  in templates

```
  - algorithms: [PERSONA_FEED]
    # heroAlgorithms: [CAMPAIGN_ITEMS]
    type: HeroListFeature
    subtitle: Hero List Feature
    title: The start of something
```
### What automated tests did you write?

coverage over api/client

## TODO:

- [x] PR has a relevant title that will be understandable in a public changelog (ie...non developers)
- [ ] Closes [tag-issues-here]
- [x] No new warnings in tests, in storybook, and in-app
- [x] Upload GIF(s) of iOS and Android
- [x] Set a relevant reviewer

## REVIEW:

**Manual QA**

- [ ] Manual QA on iOS and ensure it looks/behaves as expected
- [ ] Manual QA on Android and ensure it looks/behaves as expected

**Code Review: Questions to consider**

- [ ] Read through the "Files changed" tab _very carefully_
- [ ] Edge cases: what assumptions are made about input?
- [ ] What kind of tests could be written?
- [ ] How might we make this easier for someone else to understand?
- [ ] Could the code be simpler?
- [ ] Will the code be easy to modify in the future?
- [ ] What's one part of these changes that makes you excited to merge it?

> The purpose of PR Review is to _improve the quality of the software._

### CODE REVIEW

**Comment by vinnyjth** [comment]
@burkeshartsis I'm still wrapping up some details on this PR (loading states, tests) but it's here if you want to start reviewing it.

File: packages/apollos-data-connector-rock/src/features/data-source.js [hunk 1]
```diff
@@ -66,6 +66,55 @@ export default class Feature extends RockApolloDataSource {
...
+    if (heroAlgorithms && heroAlgorithms.length) {
+      // The actions come from the action algorithms
+      actions = () => this.runAlgorithms({ algorithms });
+      // and the hero comes from the hero algorithms
+      hero = async () => {
+        const cards = await this.runAlgorithms({ algorithms: heroAlgorithms });
+        return cards.length ? cards[0] : null;
+      };
+      // Otherwise, if we don't have a strategy
+    } else {
```
Comment (conrad-vanl) [inline packages/apollos-data-connector-rock/src/features/data-source.js:88]: Why don't we just make `heroAlgorithms` a required prop? If you wanted to use the same algorithm for both, couldn't you just do something like:

```yaml
  - title: You are my hereo
    type: HeroList
    algorithms:
      - type: CONTENT_CHANNEL
        arguments:
          skip: 1
          contentChannelId: 13
    heroAlgorithms:
      - type: CONTENT_CHANNEL
        arguments:
          contentChannelId: 13
```

File: packages/apollos-ui-connected/src/HeroListFeatureConnected/HeroListFeature/index.js [hunk 2]
```diff
@@ -92,50 +93,88 @@ const loadingStateArray = [
   },
 ];
 
+// TODO: Conrad is making this reusuable.
```
Comment (conrad-vanl) [inline packages/apollos-ui-connected/src/HeroListFeatureConnected/HeroListFeature/index.js:96]: 😬 So I imagine this PR is dependent on #1443

File: packages/apollos-data-connector-rock/src/features/data-source.js (hunk 1 above)
Comment (conrad-vanl) [reply packages/apollos-data-connector-rock/src/features/data-source.js:88]: FWIW, I think having `skip` and `limit` props on all of our core algorithms would be pretty trick and super handy...and I'd rather rely on this then having a branching conditional like this.

File: packages/apollos-ui-connected/src/HeroListFeatureConnected/HeroListFeature/index.js (hunk 2 above)
Comment (vinnyjth) [reply packages/apollos-ui-connected/src/HeroListFeatureConnected/HeroListFeature/index.js:96]: Whatever PR gets merged second will need to update ;)

File: packages/apollos-data-connector-rock/src/features/data-source.js (hunk 1 above)
Comment (conrad-vanl) [reply packages/apollos-data-connector-rock/src/features/data-source.js:88]: Buuuuut.... guess I'm also game with keeping this PR like it is and still adding `limit` and `skip` props on all of our core algo's

Comment (vinnyjth) [reply packages/apollos-data-connector-rock/src/features/data-source.js:88]: > Buuuuut.... guess I'm also game with keeping this PR like it is and still adding limit and skip props on all of our core algo's

I totally agree. We can start here and then move towards that goal. Right now none of them support `skip` and only a few of them support `limit`. I would like the branching logic to be "if no heroAlgorithm, use the normal algorithm twice with the correct skip" so you don't have to the extra configuration.

**Comment by conrad-vanl** [comment]
I'd really like to get @burkeshartsis's blessing on this from a code/architecture perspective before I give my thumbsup. It looks good from my perspective, but I ahven't been close to this as y'all have,.

File: packages/apollos-ui-connected/src/HeroListFeatureConnected/HeroListFeature/index.js [hunk 3]
```diff
@@ -92,50 +93,88 @@ const loadingStateArray = [
...
+    );
+  }
 );
 
 HeroListFeature.displayName = 'Features';
 
 HeroListFeature.propTypes = {
   // TODO: refactor ActionListCard to safely render without an actions array.
   actions: PropTypes.arrayOf(PropTypes.shape({})).isRequired, // at least for the time being this is required
+  hero: PropTypes.shape({}).isRequired, // at least for the time being this is required
```
Comment (burkeshartsis) [inline packages/apollos-ui-connected/src/HeroListFeatureConnected/HeroListFeature/index.js:177]: If we are going to keep the `action/actions` language I think we should embrace it here. `hero` feels too vague. Is it data? Is it a component? Is it a marker/boolean? Maybe call this `heroAction/heroActions`?

File: packages/apollos-ui-kit/src/ActionList/ActionListItem/index.js [hunk 4]
```diff
@@ -38,7 +38,7 @@ const CellImage = styled(({ theme }) => ({
 // eslint-disable-next-line react/prop-types
 const RenderAsTouchable = ({ children, onPress }) =>
   onPress ? (
-    <TouchableScale onPress={() => onPress}>{children}</TouchableScale>
+    <TouchableScale onPress={onPress}>{children}</TouchableScale>
```
Comment (burkeshartsis) [inline packages/apollos-ui-kit/src/ActionList/ActionListItem/index.js:41]: 🙌

File: packages/apollos-ui-connected/src/HeroListFeatureConnected/HeroListFeature/index.js (hunk 2 above)
Comment (burkeshartsis) [reply packages/apollos-ui-connected/src/HeroListFeatureConnected/HeroListFeature/index.js:96]: Is there a reason this code can't just live inline below? I feel like we don't even need `HeroItemComponent`.

File: packages/apollos-ui-connected/src/HeroListFeatureConnected/HeroListFeature/index.js (hunk 3 above)
Comment (vinnyjth) [reply packages/apollos-ui-connected/src/HeroListFeatureConnected/HeroListFeature/index.js:177]: We can call it `heroCard`

File: packages/apollos-ui-connected/src/HeroListFeatureConnected/HeroListFeature/index.js (hunk 2 above)
Comment (vinnyjth) [reply packages/apollos-ui-connected/src/HeroListFeatureConnected/HeroListFeature/index.js:96]: We can inline this or solve it permanently when #1443 is merged. I think we solve this problem the same way in the other two Features that deal with cards.

**Comment by vinnyjth** [comment]
> The main thing I'm not seeing that I had the biggest question about was how do we decide what component to render for the HeroComponent and how is that driven by the API? 

Is this a goal? I figured this card would always display the same type of component except if it was overridden globally (by passing a different `HeroComponent`. 

If we wanted to enable different component types to render, I will drag in the `cardComponentMapper` and add `isHeroFeatured` to the API, pretty much identically to how we handl the `VericalCardList` feature

**Comment by conrad-vanl** [comment]
> if being able to drive the rendered Hero card from the API is a desired feature at this time

No. None of the other features do this. I don't think we should add that to this component at this time. Let #1443 drive this for now and if it starts becoming a pressure point we can re-address

**Comment by vinnyjth** [comment]
I plan on renaming the `hero` property to `heroCard`, and then let me know if there are any other changes that should be made / thought through.

**Comment by vinnyjth** [comment]
PR updated

**Comment by burkeshartsis** [comment]
> I plan on renaming the hero property to heroCard

Are we confident the `HeroComponent` is always a card?

**Comment by vinnyjth** [comment]
> Are we confident the HeroComponent is always a card?
Yes, that's its GQL type.

**Pull request merged**